### ISSUE TITLE
HttpGet method never returns. Hangs program.

### ISSUE THREAD

**Issue by HadiSDev** [aspnet/AspNetCore#15202]
### Describe the bug
When I have a HttpGet method that is supposed to return a list of DTO's it hangs after crossing the return statement. I put a breakpoint before the return statement to confirm that there are no exceptions thrown and the list from the database is returned correctly.

### To Reproduce
1. Using this version of ASP.NET Core '3.0'
2. This is my code (Simplified):
I am using:
- EFCORE 3.0
- Asp.net NewtonsoftJson

Controller: 
```C#
[HttpGet("/somethings/GetSomething")]
        public async Task<ActionResult<IEnumerable<SomeObject>>> GetSomething()
        { 
            var result = await _repository.GetSomethingFromDB().ToListAsync();
            return result;
        }
Repository:
        public IQueryable<SomeObject> GetSomethingFromDB()
        {
            var res = _context.Somethings.Include(A => A.B).ThenInclude(B => B.H)
                                          .Include(A => A.B).ThenInclude(B=> B.L)
                                          .Include(A => A.B).ThenInclude(B =>B.Q)
                                          .Select(A => new SomeObject(A));

            return res ;
        }
```
The program does not hang if I remove ".Include(A => A.B).ThenInclude(B =>B.Q)". 
But note that GetSomethingFromDB does, in fact, return the correct result both times. Program hangs after executing "return result;" statement in the Http method

### Expected behavior
Expected to return the list without hanging.

### Screenshots

###Errors
EDIT: After some time an: An unhandled exception of type 'System.StackOverflowException' occurred in System.Private.CoreLib.dll

### Additional context
Include the output of `dotnet --info`

**Comment by javiercn** [comment]
@HadiSDev Thanks for contacting us.

Could you debug the process, break after the return and capture the callstack so that we can investigate the issue?

**Comment by pranavkm** [comment]
> EDIT: After some time an: An unhandled exception of type 'System.StackOverflowException' occurred in System.Private.CoreLib.dll

@ajcvickers does this look familiar at all?

**Comment by HadiSDev** [comment]
@javiercn Actually, the issue no longer appears. I removed AddNewtonsoftJson from AddMvc because I had a suspicion that it was causing the StackOverflow when serializing. After removing it there was no "hanging". BUT today I tried to add NewtonsoftJson again to the service builder so I could give you the debug log, and this time the problem did not reappear. Very strange. It definitely looked like an infinite loop while serializing by Json, by the way, the program behaved. 

So in short, I can't give you the log you are asking for :). I guess we will just have to wait until the problem reappears, and then I will write back if it does

**Comment by pranavkm** [comment]
Thanks @HadiSDev. MVC does not configure this value, so it should default to `Error`, but you could get recursions in your JSON model depending on the value configured in https://www.newtonsoft.com/json/help/html/P_Newtonsoft_Json_JsonSerializerSettings_ReferenceLoopHandling.htm. Yet another option might be to try and map your EF model to a POCO type and serialize it. Perhaps you may discover a cycle that causes the original error.

Since there's nothing actionable at this point, we'll close this issue. Feel free to reply to this thread logs or file a new issue once you have a consistent way to reproduce the error.

**Comment by HadiSDev** [comment]
@pranavkm well it reappeared today. Regardless if I use Newtons or not. I already map entity objects to POCO as DTO's:

Log:

Microsoft.EntityFrameworkCore.Database.Command: Information: Executing DbCommand [Parameters=[], CommandType='Text', CommandTimeout='30']
SELECT [l].[LeagueId], [l].[AlternateLeagueName], [l].[CountryCode], [l].[Divison], [l].[Gender], [l].[LeagueBadge], [l].[LeagueLogo], [l].[LeagueName], [l].[Sport], [t1].[MatchId], [t1].[AwayScore], [t1].[AwayTeamId], [t1].[HomeScore], [t1].[HomeTeamId], [t1].[LeagueId], [t1].[MatchDate], [t1].[MatchName], [t1].[Season], [t1].[Time], [t1].[TeamId], [t1].[CountryCode], [t1].[Gender], [t1].[LeagueId0], [t1].[Sport], [t1].[Stadium], [t1].[TeamBadge], [t1].[TeamName], [t1].[TeamNameShort], [t1].[TeamId0], [t1].[CountryCode0], [t1].[Gender0], [t1].[LeagueId1], [t1].[Sport0], [t1].[Stadium0], [t1].[TeamBadge0], [t1].[TeamName0], [t1].[TeamNameShort0], [t1].[BetId], [t1].[BettingType], [t1].[MatchId0], [t1].[Odds], [t1].[OppositeOdds], [t1].[Pool]
FROM [Leagues] AS [l]
LEFT JOIN (
    SELECT [m].[MatchId], [m].[AwayScore], [m].[AwayTeamId], [m].[HomeScore], [m].[HomeTeamId], [m].[LeagueId], [m].[MatchDate], [m].[MatchName], [m].[Season], [m].[Time], [t].[TeamId], [t].[CountryCode], [t].[Gender], [t].[LeagueId] AS [LeagueId0], [t].[Sport], [t].[Stadium], [t].[TeamBadge], [t].[TeamName], [t].[TeamNameShort], [t0].[TeamId] AS [TeamId0], [t0].[CountryCode] AS [CountryCode0], [t0].[Gender] AS [Gender0], [t0].[LeagueId] AS [LeagueId1], [t0].[Sport] AS [Sport0], [t0].[Stadium] AS [Stadium0], [t0].[TeamBadge] AS [TeamBadge0], [t0].[TeamName] AS [TeamName0], [t0].[TeamNameShort] AS [TeamNameShort0], [b].[BetId], [b].[BettingType], [b].[MatchId] AS [MatchId0], [b].[Odds], [b].[OppositeOdds], [b].[Pool]
    FROM [Matches] AS [m]
    LEFT JOIN [Teams] AS [t] ON [m].[HomeTeamId] = [t].[TeamId]
    LEFT JOIN [Teams] AS [t0] ON [m].[AwayTeamId] = [t0].[TeamId]
    LEFT JOIN [Bets] AS [b] ON [m].[MatchId] = [b].[MatchId]
    WHERE GETDATE() <= [m].[MatchDate]
) AS [t1] ON [l].[LeagueId] = [t1].[LeagueId]
WHERE [l].[LeagueId] IN (4328, 4331, 4332, 4334, 4335, 4340, 4344, 4480, 4481)
ORDER BY [l].[LeagueId], [t1].[MatchId], [t1].[BetId]
The thread 0x647c has exited with code 0 (0x0).
The thread 0x9398 has exited with code 0 (0x0).
The thread 0x9068 has exited with code 0 (0x0).
Microsoft.EntityFrameworkCore.Database.Command: Error: Failed executing DbCommand (30,092ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
SELECT [l].[LeagueId], [l].[AlternateLeagueName], [l].[CountryCode], [l].[Divison], [l].[Gender], [l].[LeagueBadge], [l].[LeagueLogo], [l].[LeagueName], [l].[Sport], [t1].[MatchId], [t1].[AwayScore], [t1].[AwayTeamId], [t1].[HomeScore], [t1].[HomeTeamId], [t1].[LeagueId], [t1].[MatchDate], [t1].[MatchName], [t1].[Season], [t1].[Time], [t1].[TeamId], [t1].[CountryCode], [t1].[Gender], [t1].[LeagueId0], [t1].[Sport], [t1].[Stadium], [t1].[TeamBadge], [t1].[TeamName], [t1].[TeamNameShort], [t1].[TeamId0], [t1].[CountryCode0], [t1].[Gender0], [t1].[LeagueId1], [t1].[Sport0], [t1].[Stadium0], [t1].[TeamBadge0], [t1].[TeamName0], [t1].[TeamNameShort0], [t1].[BetId], [t1].[BettingType], [t1].[MatchId0], [t1].[Odds], [t1].[OppositeOdds], [t1].[Pool]
FROM [Leagues] AS [l]
LEFT JOIN (
    SELECT [m].[MatchId], [m].[AwayScore], [m].[AwayTeamId], [m].[HomeScore], [m].[HomeTeamId], [m].[LeagueId], [m].[MatchDate], [m].[MatchName], [m].[Season], [m].[Time], [t].[TeamId], [t].[CountryCode], [t].[Gender], [t].[LeagueId] AS [LeagueId0], [t].[Sport], [t].[Stadium], [t].[TeamBadge], [t].[TeamName], [t].[TeamNameShort], [t0].[TeamId] AS [TeamId0], [t0].[CountryCode] AS [CountryCode0], [t0].[Gender] AS [Gender0], [t0].[LeagueId] AS [LeagueId1], [t0].[Sport] AS [Sport0], [t0].[Stadium] AS [Stadium0], [t0].[TeamBadge] AS [TeamBadge0], [t0].[TeamName] AS [TeamName0], [t0].[TeamNameShort] AS [TeamNameShort0], [b].[BetId], [b].[BettingType], [b].[MatchId] AS [MatchId0], [b].[Odds], [b].[OppositeOdds], [b].[Pool]
    FROM [Matches] AS [m]
    LEFT JOIN [Teams] AS [t] ON [m].[HomeTeamId] = [t].[TeamId]
    LEFT JOIN [Teams] AS [t0] ON [m].[AwayTeamId] = [t0].[TeamId]
    LEFT JOIN [Bets] AS [b] ON [m].[MatchId] = [b].[MatchId]
    WHERE GETDATE() <= [m].[MatchDate]
) AS [t1] ON [l].[LeagueId] = [t1].[LeagueId]
WHERE [l].[LeagueId] IN (4328, 4331, 4332, 4334, 4335, 4340, 4344, 4480, 4481)
ORDER BY [l].[LeagueId], [t1].[MatchId], [t1].[BetId]
'iisexpress.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\3.0.0\System.Diagnostics.StackTrace.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
'iisexpress.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\3.0.0\System.Reflection.Metadata.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
Microsoft.EntityFrameworkCore.Query: Error: An exception occurred while iterating over the results of a query for context type 'Entities.Context.BettingContext'.
Microsoft.Data.SqlClient.SqlException (0x80131904): Execution Timeout Expired.  The timeout period elapsed prior to completion of the operation or the server is not responding.
 ---> System.ComponentModel.Win32Exception (258): The wait operation timed out.
   at Microsoft.Data.SqlClient.SqlCommand.<>c.<ExecuteDbDataReaderAsync>b__164_0(Task`1 result)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__274_0(Object obj)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.AsyncQueryingEnumerable`1.AsyncEnumerator.MoveNextAsync()
ClientConnectionId:22d96fb4-b3a0-4eaf-92d1-09fe9b4d720c
Error Number:-2,State:0,Class:11

Microsoft.Data.SqlClient.SqlException (0x80131904): Execution Timeout Expired.  The timeout period elapsed prior to completion of the operation or the server is not responding.
 ---> System.ComponentModel.Win32Exception (258): The wait operation timed out.
   at Microsoft.Data.SqlClient.SqlCommand.<>c.<ExecuteDbDataReaderAsync>b__164_0(Task`1 result)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__274_0(Object obj)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.AsyncQueryingEnumerable`1.AsyncEnumerator.MoveNextAsync()
ClientConnectionId:22d96fb4-b3a0-4eaf-92d1-09fe9b4d720c
Error Number:-2,State:0,Class:11
Exception thrown: 'Microsoft.Data.SqlClient.SqlException' in System.Private.CoreLib.dll
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker: Information: Executed action WebAPI.Controllers.Public.LeaguesController.GetAvailableLeagues (BettingWeb) in 33012.0631ms
Microsoft.AspNetCore.Routing.EndpointMiddleware: Information: Executed endpoint 'WebAPI.Controllers.Public.LeaguesController.GetAvailableLeagues (BettingWeb)'
Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware: Error: An unhandled exception has occurred while executing the request.

Microsoft.Data.SqlClient.SqlException (0x80131904): Execution Timeout Expired.  The timeout period elapsed prior to completion of the operation or the server is not responding.
 ---> System.ComponentModel.Win32Exception (258): The wait operation timed out.
   at Microsoft.Data.SqlClient.SqlCommand.<>c.<ExecuteDbDataReaderAsync>b__164_0(Task`1 result)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__274_0(Object obj)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReaderAsync(RelationalCommandParameterObject parameterObject, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.AsyncQueryingEnumerable`1.AsyncEnumerator.MoveNextAsync()
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
   at WebAPI.Controllers.Public.LeaguesController.GetAvailableLeagues() in G:\Hubly\bettingASP\BettingSite\BettingWeb\Controllers\Public\LeaguesController.cs:line 27
   at lambda_method(Closure , Object )
   at Microsoft.Extensions.Internal.ObjectMethodExecutorAwaitable.Awaiter.GetResult()
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)
   at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
ClientConnectionId:22d96fb4-b3a0-4eaf-92d1-09fe9b4d720c
Error Number:-2,State:0,Class:11
Microsoft.AspNetCore.Hosting.Diagnostics: Information: Request finished in 33188.922ms 500 text/plain

**Comment by HadiSDev** [comment]
@pranavkm I should also mention that it sometimes works and sometimes does not. For most it does not and even when it does it is very slow. I am running on LocalDB MSSQL. I tried to execute the query in MSSQL studio manually, and it runs without any problems. It is hard for me to judge where if EF core is to blame, myself or ASP.NET Core 3.0

**Comment by javiercn** [comment]
@HadiSDev 

Could you call ToList() and put the result in a local variable before returning it, then put a break point on the `return result` statement and see if you get to hit it?

Does it run past the ToList() call?

**Comment by HadiSDev** [comment]
@javiercn first 4 times I tried what you asked me to do, it did not hit on return and threw the error. After that, I would get a hit every time on return and no Hanging nor errors. Restarted several times to test again, and it hit every time.

**Comment by javiercn** [comment]
@ajcvickers this seems to be an issue in EF, can you get someone to take a look?
